garage/envs/base.py

```diff
@@ -1,17 +1,138 @@
+"""Wrapper class that converts gym.Env into GarageEnv."""
 import collections
+import warnings
 
-from cached_property import cached_property
+import gym
+from gym.spaces import Box as GymBox
+from gym.spaces import Discrete as GymDiscrete
+from gym.spaces import Tuple as GymTuple
 
-from garage.envs import EnvSpec
+from garage.core import Parameterized
+from garage.core import Serializable
+from garage.envs.env_spec import EnvSpec
+from garage.spaces import Box
+from garage.spaces import Discrete
+from garage.spaces import Product
 
-_Step = collections.namedtuple("Step",
-                               ["observation", "reward", "done", "info"])
 
+class GarageEnv(gym.Wrapper, Parameterized, Serializable):
+    """
+    Returns a Garage wrapper class for gym.Env.
+
+    In order to provide pickling (serialization) and parameterization
+    for gym.Envs, they must be wrapped with GarageEnv. This ensures
+    compatibility with existing samplers and checkpointing when the
+    envs are passed internally around garage.
 
-def Step(observation, reward, done, **kwargs):
+    Furthermore, GarageEnv silently converts action_space and
+    observation_space from gym.Spaces to garage.spaces.
+
+    Args:
+        env (gym.Env): the env that will be wrapped
     """
-    Convenience method creating a namedtuple with the results of the
-    environment.step method.
-    Put extra diagnostic info in the kwargs
+
+    def __init__(self, env):
+        Serializable.quick_init(self, locals())
+        Parameterized.__init__(self)
+        super().__init__(env)
+        self.action_space = self._to_garage_space(env.action_space)
```

Comment:
Aren't these properties in TfEnv/TheanoEnv? Why set a member var here?

It seems better not to set them there--then if the property doesn't exists (i.e. someone constructed a GarageEnv outside of a TfEnv/TheanoEnv) we will get an error.

Reply:
http://sayspy.blogspot.com/2009/12/how-to-handle-multiple-inheritance-of.html

Reply:
The issue here is wherever I call `self.action_space` (e.g. in `spec()`), it defaults to gym.spaces.Box, even if I have @overrides above `def action_space(...)` in TheanoEnv/TfEnv.

Reply:
As in even if the property `def action_space` does exist, calling `self.action_space` will just default to the underlying env.action_space (which is a gym space), instead of returning the property method.